### ISSUE TITLE
Backwards incompatible API changes to setIdentity, setVersion

### ISSUE THREAD

**Issue by stebbi** [orientechnologies/orientdb#3254]
Hi, 

I'm using OrientDB as a new data store for an existing application. I frequently (1) query for one or more ODocument objects, (2) convert the objects to an alternate representation expected by a legacy API, (3) pass the converted objects to the legacy API and get new objects with changes, and (4) synthesize new ODocument objects to write back to the store. 

When I synthesize the ODocument objects, I use setIdentity and setVersion to ensure that OrientDB correctly updates the objects I was working with, so they're stored using the old RIDs and new version numbers. Otherwise I'd wind up with copies, where each update would create a new RID. 

All of this was fine with OrientDB 1.7.\* but no longer works with 2.0-rc1. The ORecordAbstract.setIdentity and ORecordAbstract.setVersion methods are now protected and my codebase basically breaks. 

I'd appreciate if the methods could be made public again, or if not, then there should be a workaround for the case I have described, where OrientDB-backed objects have to be processed by libraries with APIs that take something other than e.g. ODocument & friends.

**Comment by buzzkillington** [comment]
I don't understand why there shouldn't be API support for fully constructing ODocuments, including the version, as the OP described. Consider the use case of a web-based API built on top of OrientDB, in which multiple clients are retrieving/updating objects (which include a serialized version number).  If an update is received through the API, shouldn't it be possible to use Orient's version checking by setting the version on an ODocument constructed to represent that update (to ensure that it's not based on an old version, and potentially overwriting new data)?

**Comment by stebbi** [comment]
This is basically the same scenario I was working with - I think exactly
the same, from the OrientDB API viewpoint.

**Comment by buzzkillington** [comment]
@lvca / @tglman - can we get this issue (or at least the discussion) re-opened?

**Comment by stebbi** [comment]
How about ORecordInternal? I haven't done anything about this yet but that
sounded like a solution to me.

**Comment by buzzkillington** [comment]
I disagree with that answer being a solution: using an undocumented/unsupported, internal mechanism is (at best) a kludgy, unsatisfactory workaround.  Regarding the other option described above, it seems unnecessary and unreasonable to incur two database hits and perform a manual, client-side version check just to process an update.  This should be a public API-supported capability - otherwise, I'd like a more well-reasoned explanation for why, given our use cases, "normal application flow" is defined so narrowly (comment from @tglman above).

**Comment by PhantomYdn** [comment]
Give me a try to explain.
The logic is following: it's good for overall performance and memory consumption to have just 1 instance for a document in the memory per transaction. If you will try to load/save a document with described way you might have theoretically have 2 java objects for the same document in the memory. And that's might lead to a sequence of problems.

So let me back my previous comments: I think it's definetly should be done in that way as OT guys suggest. And more over: there is almost no performance or usability impact for the client side if you will use methods merge().

**Comment by stebbi** [comment]
Hmm - how about that then, @tlgman?

I do have to admit that I can't quite follow Ilya's argument. And I did
file this as a bug whereas now we're talking about a feature. But either
way, perhaps the generalized scenario of receiving every object field value
from a JSON API, HTTP push or some equivalent, and blindly stuffing in
OrientDB as fastcas possible, is worth keeping around as an option?

I think perhaps the main reason I won't be as decisive on this as
@buzzkillington is that I remember what a painful couple of bugs I ran into
with mismatched version numbers - no fun to resolve...
